osu.Game/Screens/Edit/HitObjectChangeHandler.cs

```diff
@@ -0,0 +1,293 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using osu.Framework.Bindables;
+using osu.Framework.Logging;
+using osu.Game.Rulesets.Objects;
+using osu.Game.Screens.Edit.Changes;
+
+namespace osu.Game.Screens.Edit
+{
+    /// <summary>
+    /// A change handler for handling hit object changes in an <see cref="EditorBeatmap"/> using the command design pattern.
+    /// This is supposed to eventually replace the <see cref="EditorChangeHandler"/> and its inheritors once all the editor operations have been refactored.
+    /// </summary>
+    /// <inheritdoc/>
+    public partial class HitObjectChangeHandler : TransactionalCommitComponent
+    {
+        private readonly EditorBeatmap editorBeatmap;
+
+        /// <summary>
+        /// This change handler will be suppressed while a transaction with this command handler is in progress.
+        /// Any save states of this change handler will be added to the undo stack.
+        /// </summary>
+        private readonly EditorChangeHandler? changeHandler;
+
+        public readonly Bindable<bool> CanUndo = new BindableBool();
+
+        public readonly Bindable<bool> CanRedo = new BindableBool();
+
+        public event Action? OnStateChange;
+
+        public bool HasUncommittedChanges => currentTransaction.UndoChanges.Count != 0;
+
+        public Guid CurrentState => undoStack.TryPeek(out var transaction) ? transaction.Id : Guid.Empty;
+
+        private Transaction currentTransaction;
+
+        private readonly Stack<Transaction> undoStack = new Stack<Transaction>();
+
+        private readonly Stack<Transaction> redoStack = new Stack<Transaction>();
+
+        private bool isRestoring;
+
+        public HitObjectChangeHandler(EditorBeatmap editorBeatmap, EditorChangeHandler? changeHandler = null)
+        {
+            currentTransaction = new Transaction();
+            this.editorBeatmap = editorBeatmap;
+            this.changeHandler = changeHandler;
+
+            editorBeatmap.TransactionBegan += BeginChange;
+            editorBeatmap.TransactionEnded += EndChange;
+            editorBeatmap.SaveStateTriggered += SaveState;
+
+            if (this.changeHandler != null)
+                this.changeHandler.OnStateChange += commitChangeHandlerStateChange;
+        }
+
+        /// <summary>
+        /// Submits a change to be applied and added to the history.
+        /// </summary>
+        /// <param name="change">Change to be applied.</param>
+        /// <param name="commitImmediately">Whether to commit the current transaction and push it onto the undo stack immediately.</param>
+        public void Submit(IRevertibleChange change, bool commitImmediately = false)
```

Comment:
Why does `commitImmediately` exist and why does it default to false?

I presume it's there for nesting commands. If that is accurate, then in order of preference:

- it should not exist, and `BeginChange()` / `EndChange()` should support nesting as it already does in the current system, obviating the need for this (if you're in a nested change `BeginChange()` is essentially a no-op)
- it should default to true because that's the obvious use